### DISCUSSION TITLE
How to get generated routes?

### DISCUSSION THREAD

**Discussion by nirtamir2** [solidjs/solid-start#588] [Q&A]
Hi
I want to create a small adapter for solid-start in [routes-gen](https://github.com/sandulat/routes-gen).
This will allow using type-safe routes for solid-start.
I could replicate the solid routes resolution manually, but after looking at the remix adapter - it seems that in order to get the routes we just need to run
`remix routes --json` . From there it's easy to parse them.
In solid-start we have the
- Private `/dev/print-routes.js` that print the routes in dev & build, and log to the terminal  output, but we don't know when it finishes
- Private `router.getFlattenedApiRoutes(showPageRoutes: boolean)` 
- The build result file `.solid/inspect/route-manifest.json` which we can infer the page routes from it keys (I could build in a temp folder and take it but I don't have a way to do it without config change and its not efficient)

What should be the best way to handle it? Can we expose `solid-start routes --json` too?, or its better that I will handle it manually (which is ok for me too)

**Comment by nirtamir2** [comment]
We can add this in `bin.cjs`
```ts
prog
  .command("routes").describe("print routes")
  .action(async ({config: configFile, open, port, root, host, inspect}) => {
    root = root || process.cwd();
    const config = await resolveConfig({ mode: "production", configFile, root, command: "build" });
    
    const { Router } = await import("./fs-router/router.js");

    const router = new Router({
      baseDir: path.posix.join(config.solidOptions.appRoot, config.solidOptions.routesDir),
      pageExtensions: config.solidOptions.pageExtensions,
      ignore: config.solidOptions.routesIgnore,
      cwd: config.solidOptions.root
    });
    await router.init();

    console.log(JSON.stringify(router.getFlattenedPageRoutes(), null, 2));
  });
```
to get the desired output.


<details>
<summary>Running solid-start routes</summary>

`> solid-start routes --json`
```json

[
  {
    "id": "/(a)",
    "path": "",
    "componentPath": "src/routes/(a).tsx"
  },
  {
    "id": "/[...404]",
    "path": "/*404",
    "componentPath": "src/routes/[...404].tsx"
  },
  {
    "id": "/[param]",
    "path": "/:param",
    "componentPath": "src/routes/[param].tsx"
  },
  {
    "id": "/index",
    "path": "/",
    "componentPath": "src/routes/index.tsx"
  }
]

```

</details>





<details>
<summary>Running remix routes --json</summary>
`> remix routes --json`
```json

[
  {
    "id": "root",
    "path": "",
    "file": "root.tsx",
    "children": [
      {
        "id": "routes/workshops/$workshopId.edit",
        "path": "workshops/:workshopId/edit",
        "file": "routes/workshops/$workshopId.edit.tsx"
      },
      {
        "id": "routes/workshops/$workshopId",
        "path": "workshops/:workshopId",
        "file": "routes/workshops/$workshopId.tsx"
      },
      {
        "id": "routes/$test2",
        "path": ":test2",
        "file": "routes/$test2.tsx"
      },
      {
        "id": "routes/(test)",
        "path": "(test)",
        "file": "routes/(test).tsx"
      },
      {
        "id": "routes/logout",
        "path": "logout",
        "file": "routes/logout.tsx"
      },
      {
        "id": "routes/index",
        "index": true,
        "file": "routes/index.tsx"
      },
      {
        "id": "routes/login",
        "path": "login",
        "file": "routes/login.tsx"
      },
      {
        "id": "routes/test",
        "path": "test",
        "file": "routes/test.tsx"
      }
    ]
  }
]


```



</details>




<details>
<summary>Details</summary>



</details>

**Comment by ryansolid** [comment]
Yeah if this gets what you are looking for seems reasonable to me.

**Reply by nirtamir2** [reply]
Great! I will work on it this weekend :)

**Reply by nirtamir2** [reply]
I created a PR - https://github.com/solidjs/solid-start/pull/609
I will help me doing creating plugin to routes-gen https://github.com/sandulat/routes-gen/pull/29
